src/type-factor.c

```diff
@@ -74,3 +74,287 @@ static SEXP levels_union(SEXP x, SEXP y) {
   UNPROTECT(2);
   return out;
 }
+
+// -----------------------------------------------------------------------------
+
+static void new_factor(SEXP x, SEXP levels);
+static void new_ordered(SEXP x, SEXP levels);
+
+
+// [[ include("vctrs.h") ]]
+SEXP fct_as_character(SEXP x, struct vctrs_arg* x_arg) {
+  SEXP levels = Rf_getAttrib(x, R_LevelsSymbol);
+
+  if (TYPEOF(levels) != STRSXP) {
+    stop_corrupt_factor_levels(x, x_arg);
+  }
+
+  return Rf_asCharacterFactor(x);
+}
+
+// [[ include("vctrs.h") ]]
+SEXP ord_as_character(SEXP x, struct vctrs_arg* x_arg) {
+  return fct_as_character(x, x_arg);
+}
+
+
+static SEXP chr_as_factor_from_self(SEXP x, bool ordered);
+static SEXP chr_as_factor_impl(SEXP x, SEXP levels, bool* lossy, bool ordered);
+
+// [[ include("vctrs.h") ]]
+SEXP chr_as_factor(SEXP x, SEXP to, bool* lossy, struct vctrs_arg* to_arg) {
+  SEXP levels = Rf_getAttrib(to, R_LevelsSymbol);
+
+  if (TYPEOF(levels) != STRSXP) {
+    stop_corrupt_factor_levels(to, to_arg);
+  }
+
+  // When `to` has no levels, it is treated as a template and the
+  // levels come from `x`
+  if (vec_size(levels) == 0) {
+    return chr_as_factor_from_self(x, false);
+  }
+
+  return chr_as_factor_impl(x, levels, lossy, false);
+}
+
+// [[ include("vctrs.h") ]]
+SEXP chr_as_ordered(SEXP x, SEXP to, bool* lossy, struct vctrs_arg* to_arg) {
+  SEXP levels = Rf_getAttrib(to, R_LevelsSymbol);
+
+  if (TYPEOF(levels) != STRSXP) {
+    stop_corrupt_ordered_levels(to, to_arg);
+  }
+
+  // When `to` has no levels, it is treated as a template and the
+  // levels come from `x`
+  if (vec_size(levels) == 0) {
+    return chr_as_factor_from_self(x, true);
+  }
+
+  return chr_as_factor_impl(x, levels, lossy, true);
+}
+
+static SEXP chr_as_factor_impl(SEXP x, SEXP levels, bool* lossy, bool ordered) {
+  SEXP out = PROTECT(vec_match(x, levels));
+  const int* p_out = INTEGER(out);
+
+  R_len_t size = vec_size(x);
+  const SEXP* p_x = STRING_PTR_RO(x);
+
+  // Detect lossy no-matches, but allow `NA` values from `x`
+  for (R_len_t i = 0; i < size; ++i) {
+    if (p_out[i] != NA_INTEGER) {
```

Comment:
I think it'd be a little clearer if the branch was `out == NA && x != NA` with the lossy `return` inside it, and no `continue` in the loop.